0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+
+# Motivation
+[motivation]: #motivation
+
+The two primary motivations behind this change are correctness and compatibility.
+
+## Correctness
+
+The current definition of these types means that the Dafny `string` type allows data that is not a valid Unicode string.
+The value `"\uD800"`, for example, is not a valid Unicode string and has no valid encoding in UTF-8, UTF-16,
+or any other encoding.
+This means that any logic to process strings must manually specify pre-conditions to exclude invalid values, or
+be less precise with its specification. For example, an implementation of UTF-8 encoding would have to reject
+invalid inputs in a pre-condition, dynamically fail if given invalid strings, or implement a lossy encoding by
+discarding or replacing invalid characters.
+
+## Compatibility
+
+The current definitions of `string` and `char` are biased towards using a UTF-16 encoding at runtime.
+This aligns well with some compilation target languages which also use UTF-16, such as Java, C#, and JavaScript, 
+but less well with those that use the recently more popular UTF-8 encoding, such as Go or Rust.
+Any Dafny code that interfaces with external code will often have to convert between `string` values and
+native representations of strings, and baking in the assumption of UTF-16 imposes a complexity and performance
+penalty for multiple target environments.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As of Dafny 4.0, the default semantics of the `char` type have been changed to represent Unicode scalar values instead of UTF-16 code points.
+This means that surrogate code points, meaning those in the range `[0xD800, 0xDFFF]`, are no longer allowed. This also means
+code points that require more than 16 bits, meaning those in the range `[0x10000, 0x1FFFF]`, are now directly representable
+as `char` values.
+
+The definition of the `char` type is controlled by a command-line switch called `/unicodeChar`, where `/unicodeChar:0`
+enables the previous semantics and `/unicodeChar:1` enables the new semantics described above.
+The default prior to Dafny 4.0 is `/unicodeChar:0`, but the default as of Dafny 4.0 is `/unicodeChar:1`.
+This change in definition can cause previously verified code to no longer verify,
+since expressions such as `someIntValue as char` or `someCharValue + 'a'` are no longer allowed to result in
+surrogate code points.
+
+This change can also result in behavioral changes that may not be caught by verification,
+as they may not intersect with any explicit specifications.
+Here is an example of a program that will behave differently depending on the value of `/unicodeChar`,
+and hence differently on Dafny 4.0 compared to earlier versions:
+
+```dafny
+method Main() {
+  var s := "Unicode is just so \ud83d\ude0e";
```

Comment:
This line ought to give an error with `/unicodeChar:1`, since `\ud83d` is not a legal character.

Reply:
Turns out I'm on board with that after all. :)